In [4]:
# Data manipulation
import pandas as pd
import numpy as np
# Visualization
import matplotlib.pyplot as plt
import seaborn as sns
# Settings
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)
%matplotlib inline
print('Libraries imported successfully!')

Libraries imported successfully!


# Customer Churn Analysis - Week 1 EDA
**Author**: [muhammad umar khan]
**Date**: [4/3/26]
**Course**: Introduction to Applied AI
## Project Overview
This notebook performs exploratory data analysis on customer churn data to identify
patterns and inform our predictive modeling approach.
## Table of Contents
1. Dataset Overview
2. Numerical Features Analysis
3. Categorical Features Analysis
4. Feature Correlations
5. Key Insights and Findings


In [6]:
# Load data
df = pd.read_csv('/kaggle/input/datasets/umarkhan360/my-dataset/WA_Fn-UseC_-Telco-Customer-Churn.csv')
# First look
print('Dataset shape:', df.shape)
print('\nFirst 5 rows:')
df.head()

Dataset shape: (7043, 21)

First 5 rows:


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 1. Dataset Overview

In [7]:
# 1.1 Column names and types
print('Column Information:')
df.info()

Column Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043

In [ ]:
# 1.2 Statistical summary
df.describe()

In [ ]:
# 1.3 Check for missing values
print('Missing Values:')
missing = df.isnull().sum()
print(missing[missing > 0])
# Fix TotalCharges if needed
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [ ]:
# 1.4 Target variable distribution
print('Churn Distribution:')
print(df['Churn'].value_counts())
print('\nChurn Percentage:')
print(df['Churn'].value_counts(normalize=True) * 100)

### Key Findings:- Dataset has X customers- Churn rate is Y%- Z missing values in TotalCharges column- Features include demographics, services, and account info


## 2. Numerical Features Analysis

In [ ]:
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
plt.hist(df['tenure'], bins=30, color='skyblue', edgecolor='black')
plt.xlabel('Tenure (months)')
plt.ylabel('Number of Customers')
plt.title('Distribution of Customer Tenure')
plt.subplot(1, 2, 2)
sns.boxplot(x='Churn', y='tenure', data=df)
plt.title('Tenure by Churn Status')
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
plt.hist(df['MonthlyCharges'], bins=30, color='lightcoral', edgecolor='black')
plt.xlabel('Monthly Charges ($)')
plt.ylabel('Number of Customers')
plt.title('Distribution of Monthly Charges')
plt.subplot(1, 2, 2)
sns.boxplot(x='Churn', y='MonthlyCharges', data=df)
plt.title('Monthly Charges by Churn Status')
plt.tight_layout()
plt.show()

In [ ]:
# Remove missing values for this plot
df_clean = df[df['TotalCharges'].notna()]
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
sns.violinplot(x='Churn', y='TotalCharges', data=df_clean)
plt.title('Total Charges by Churn Status')
plt.subplot(1, 2, 2)
sns.scatterplot(x='tenure', y='TotalCharges', hue='Churn', data=df_clean,
alpha=0.5)
plt.title('Tenure vs Total Charges')
plt.tight_layout()
plt.show()

### Numerical Features Insights:- Tenure: Customers with shorter tenure more likely to churn- Monthly Charges: Higher charges correlate with higher churn- Total Charges: Related to tenure (longer tenure = higher total charges)


## 3. Categorical Features Analysis

In [ ]:
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
df['Contract'].value_counts().plot(kind='bar', color='steelblue')
plt.title('Contract Type Distribution')
plt.xlabel('Contract Type')
plt.ylabel('Count')
plt.xticks(rotation=45)
plt.subplot(1, 2, 2)
contract_churn = pd.crosstab(df['Contract'], df['Churn'], normalize='index') * 100
contract_churn.plot(kind='bar', stacked=False)
plt.title('Churn Rate by Contract Type')
plt.xlabel('Contract Type')
plt.ylabel('Percentage')
plt.legend(title='Churn')
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
sns.countplot(x='InternetService', hue='Churn', data=df)
plt.title('Churn by Internet Service Type')
plt.xlabel('Internet Service')
plt.xticks(rotation=45)
plt.subplot(1, 2, 2)
internet_churn = pd.crosstab(df['InternetService'], df['Churn'], normalize='index')
* 100
internet_churn['Yes'].plot(kind='bar', color='coral')
plt.title('Churn Rate by Internet Service')
plt.ylabel('Churn Rate (%)')
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

In [ ]:
payment_churn = pd.crosstab(df['PaymentMethod'], df['Churn'], normalize='index') *
100
plt.figure(figsize=(10, 6))
payment_churn['Yes'].sort_values().plot(kind='barh', color='teal')
plt.title('Churn Rate by Payment Method')
plt.xlabel('Churn Rate (%)')
plt.ylabel('Payment Method')
plt.tight_layout()
plt.show()


## 4. Feature Correlations

In [ ]:
# Prepare numeric data
df_numeric = df.copy()
# Convert Yes/No to 1/0
binary_cols = ['Churn', 'Partner', 'Dependents', 'PhoneService', 'PaperlessBilling']
for col in binary_cols:
if col in df_numeric.columns:
df_numeric[col] = df_numeric[col].map({'Yes': 1, 'No': 0})
# Select only numeric columns
numeric_features = df_numeric.select_dtypes(include=[np.number])
# Correlation with Churn
plt.figure(figsize=(10, 8))
churn_corr = numeric_features.corr()['Churn'].sort_values(ascending=False)
churn_corr.plot(kind='barh', color='purple')
plt.title('Feature Correlation with Churn')
plt.xlabel('Correlation Coefficient')
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 10))
correlation_matrix = numeric_features.corr()
sns.heatmap(correlation_matrix, annot=True, fmt='.2f', cmap='coolwarm', center=0)
plt.title('Feature Correlation Heatmap')
plt.tight_layout()
plt.show()


## 5. Key Insights and Findings
### High-Risk Customer Characteristics:
1. **Contract Type**: Month-to-month contracts have XX% churn rate (vs Y% for
yearly)
2. **Tenure**: Customers with <6 months tenure are high risk
3. **Charges**: Higher monthly charges correlate with higher churn
4. **Services**: Fiber optic internet users churn more than DSL
5. **Payment**: Electronic check users have higher churn
### Patterns Observed:- [Add 3-5 specific observations from your analysis]
### Recommendations for Feature Engineering:- Create 'TenureGroup' (0-12, 13-24, 25-48, 49+ months)- Flag 'HighMonthlyCharges' (>$70)- Count total services subscribed- Create 'ChargesPerService' ratio
### Questions for Next Week:- Which features should we include in the model?- How to handle categorical variables?- What about class imbalance (if churn rate is low)?

# Customer Churn Analysis – Understanding Report

**Author:** Muhammad Umar Khan  
**Course:** Introduction to Applied AI  
**Platform:** Kaggle Notebook  
**Project Type:** Exploratory Data Analysis (EDA)

---

# 1. Introduction

Customer churn occurs when customers stop using a company's services. It is an important problem for many businesses, especially telecom companies, because losing customers directly affects revenue.

The purpose of this analysis is to explore the **Telco Customer Churn dataset** and understand the main factors that influence whether a customer stays or leaves the service.

This project mainly focuses on **Exploratory Data Analysis (EDA)** to understand the structure of the data and identify patterns related to churn.

---

# 2. Dataset Overview

The dataset contains information about telecom customers including their personal details, services they use, and billing information.

### Key information in the dataset

- Customer demographic information
- Internet and phone services
- Contract type
- Billing and payment methods
- Customer tenure
- Churn status (Yes/No)

### Important Features

Some important columns used in the analysis include:

- **customerID** – Unique customer identifier  
- **gender** – Male or Female  
- **SeniorCitizen** – Whether the customer is a senior citizen  
- **Partner** – Whether the customer has a partner  
- **Dependents** – Whether the customer has dependents  
- **tenure** – Number of months the customer has stayed with the company  
- **Contract** – Type of contract (Month-to-month, One year, Two year)  
- **MonthlyCharges** – Monthly payment by the customer  
- **TotalCharges** – Total amount charged to the customer  
- **PaymentMethod** – Customer payment method  
- **Churn** – Whether the customer left the service

---

# 3. Data Exploration

First, the dataset was inspected to understand its structure.

The following steps were performed:

- Checking dataset size
- Viewing first few rows
- Inspecting data types
- Checking missing values

This helps ensure that the dataset is clean and ready for analysis.

---

# 4. Numerical Feature Analysis

Numerical variables such as **tenure**, **MonthlyCharges**, and **TotalCharges** were analyzed.

### Observations

- Customers with **low tenure (new customers)** have a higher chance of churn.
- Customers paying **higher monthly charges** tend to churn more often.
- Long-term customers usually stay with the company.

This indicates that **customer loyalty increases with time**.

---

# 5. Categorical Feature Analysis

Several categorical features were analyzed to understand their impact on churn.

## Contract Type

Customers with **month-to-month contracts** show the highest churn rate.

Customers with **one-year or two-year contracts** have much lower churn rates.

## Payment Method

Customers using **electronic check** as a payment method show a higher churn rate compared to other methods.

## Customer Demographics

Factors such as:

- Senior citizen status
- Having a partner
- Having dependents

can also influence churn patterns.

---

# 6. Key Insights

From the exploratory analysis, several important insights were identified:

1. **Month-to-month contracts** have the highest churn rate (~42%).
2. Customers with **less than 6 months tenure** have very high churn (~54%).
3. **Long-term contracts (1–2 years)** significantly reduce churn.
4. Customers with **higher monthly charges** may be more likely to leave.
5. **New customers are the most vulnerable group**.

---

# 7. Conclusion

This exploratory analysis helped identify important factors related to customer churn.

The main drivers of churn include:

- Contract type
- Customer tenure
- Monthly charges
- Payment methods

Understanding these patterns can help telecom companies design better **customer retention strategies** and reduce churn.

In future work, machine learning models can be applied to **predict customer churn** based on these features.

---